File: cpp/src/platform/slug.cpp
#include <retro_metadata/platform/slug.hpp>

#include <unordered_map>

namespace retro_metadata {

namespace {

// Maps slugs to human-readable names
const std::unordered_map<std::string_view, std::string_view> kSlugNames = {
    {platform_slug::k3DO, "3DO Interactive Multiplayer"},
    {platform_slug::kAcpc, "Amstrad CPC"},
    {platform_slug::kAmiga, "Amiga"},
    {platform_slug::kAmigaCD, "Amiga CD"},
    {platform_slug::kAmigaCD32, "Amiga CD32"},
    {platform_slug::kAmstradGX4000, "Amstrad GX4000"},
    {platform_slug::kAndroid, "Android"},
    {platform_slug::kAppleII, "Apple II"},
    {platform_slug::kAppleIIGS, "Apple IIGS"},
    {platform_slug::kArcade, "Arcade"},
    {platform_slug::kAtari2600, "Atari 2600"},
    {platform_slug::kAtari5200, "Atari 5200"},
    {platform_slug::kAtari7800, "Atari 7800"},
    {platform_slug::kAtari8bit, "Atari 8-bit"},
    {platform_slug::kAtariJaguarCD, "Atari Jaguar CD"},
    {platform_slug::kAtariST, "Atari ST"},
    {platform_slug::kAtariXEGS, "Atari XEGS"},
    {platform_slug::kBBCMicro, "BBC Micro"},
    {platform_slug::kC128, "Commodore 128"},
    {platform_slug::kC16, "Commodore 16"},
    {platform_slug::kC64, "Commodore 64"},
    {platform_slug::kCPlus4, "Commodore Plus/4"},
    {platform_slug::kColecovision, "ColecoVision"},
    {platform_slug::kCommodoreCDTV, "Commodore CDTV"},
    {platform_slug::kCPS1, "CPS-1"},
    {platform_slug::kCPS2, "CPS-2"},
    {platform_slug::kCPS3, "CPS-3"},
    {platform_slug::kDC, "Sega Dreamcast"},
    {platform_slug::kDOS, "DOS"},
    {platform_slug::kEvercade, "Evercade"},
    {platform_slug::kFairchildChannelF, "Fairchild Channel F"},
    {platform_slug::kFamicom, "Famicom"},
    {platform_slug::kFDS, "Famicom Disk System"},
    {platform_slug::kFMTowns, "FM Towns"},
    {platform_slug::kGameGear, "Sega Game Gear"},
    {platform_slug::kGamate, "Gamate"},
    {platform_slug::kGameDotCom, "Game.com"},
    {platform_slug::kGB, "Game Boy"},
    {platform_slug::kGBA, "Game Boy Advance"},
    {platform_slug::kGBC, "Game Boy Color"},
    {platform_slug::kGenesis, "Sega Genesis"},
    {platform_slug::kGizmondo, "Gizmondo"},
    {platform_slug::kIOS, "iOS"},
    {platform_slug::kIntellvision, "Intellivision"},
    {platform_slug::kJaguar, "Atari Jaguar"},
    {platform_slug::kLinux, "Linux"},
    {platform_slug::kLynx, "Atari Lynx"},
    {platform_slug::kMac, "Mac"},
    {platform_slug::kMSX, "MSX"},
    {platform_slug::kMSX2, "MSX2"},
    {platform_slug::kMSX2Plus, "MSX2+"},
    {platform_slug::kN3DS, "Nintendo 3DS"},
    {platform_slug::kN64, "Nintendo 64"},
    {platform_slug::kN64DD, "Nintendo 64DD"},
    {platform_slug::kNDS, "Nintendo DS"},
    {platform_slug::kNeoGeoAES, "Neo Geo AES"},
    {platform_slug::kNeoGeoCD, "Neo Geo CD"},
    {platform_slug::kNeoGeoMVS, "Neo Geo MVS"},
    {platform_slug::kNeoGeoPocket, "Neo Geo Pocket"},
    {platform_slug::kNeoGeoPocketColor, "Neo Geo Pocket Color"},
    {platform_slug::kNES, "Nintendo Entertainment System"},
    {platform_slug::kNewNintendo3DS, "New Nintendo 3DS"},
    {platform_slug::kNGage, "N-Gage"},
    {platform_slug::kNGC, "Nintendo GameCube"},
    {platform_slug::kNintendoDSi, "Nintendo DSi"},
    {platform_slug::kOdyssey2, "Magnavox Odyssey 2"},
    {platform_slug::kOuya, "Ouya"},
    {platform_slug::kPC8800, "PC-8800 Series"},
    {platform_slug::kPC9800, "PC-9800 Series"},
    {platform_slug::kPCFX, "PC-FX"},
    {platform_slug::kPlaydate, "Playdate"},
    {platform_slug::kPocketstation, "PocketStation"},
    {platform_slug::kPokemonMini, "Pokemon Mini"},
    {platform_slug::kPS2, "PlayStation 2"},
    {platform_slug::kPS3, "PlayStation 3"},
    {platform_slug::kPS4, "PlayStation 4"},
    {platform_slug::kPS5, "PlayStation 5"},
    {platform_slug::kPSP, "PlayStation Portable"},
    {platform_slug::kPSVita, "PlayStation Vita"},
    {platform_slug::kPSVR, "PlayStation VR"},
    {platform_slug::kPSVR2, "PlayStation VR2"},
    {platform_slug::kPSX, "PlayStation"},
    {platform_slug::kSatellaview, "Satellaview"},
    {platform_slug::kSaturn, "Sega Saturn"},
    {platform_slug::kSega32, "Sega 32X"},
    {platform_slug::kSegaCD, "Sega CD"},
    {platform_slug::kSegaCD32, "Sega CD 32X"},
    {platform_slug::kSegaPico, "Sega Pico"},
    {platform_slug::kSeriesXS, "Xbox Series X|S"},
    {platform_slug::kSFam, "Super Famicom"},
    {platform_slug::kSG1000, "Sega SG-1000"},
    {platform_slug::kSharpX68000, "Sharp X68000"},
    {platform_slug::kSMS, "Sega Master System"},
    {platform_slug::kSNES, "Super Nintendo"},
    {platform_slug::kStadia, "Google Stadia"},
    {platform_slug::kSuperGrafx, "SuperGrafx"},
    {platform_slug::kSupervision, "Supervision"},
    {platform_slug::kSwitch, "Nintendo Switch"},
    {platform_slug::kTG16, "TurboGrafx-16"},
    {platform_slug::kTurboGrafxCD, "TurboGrafx-CD"},
    {platform_slug::kVectrex, "Vectrex"},
    {platform_slug::kVIC20, "VIC-20"},
    {platform_slug::kVirtualBoy, "Virtual Boy"},
    {platform_slug::kWii, "Wii"},
    {platform_slug::kWiiU, "Wii U"},
    {platform_slug::kWin, "Windows"},
    {platform_slug::kWin3x, "Windows 3.x"},
    {platform_slug::kWonderSwan, "WonderSwan"},
    {platform_slug::kWonderSwanColor, "WonderSwan Color"},
    {platform_slug::kX1, "Sharp X1"},
    {platform_slug::kXbox, "Xbox"},
    {platform_slug::kXbox360, "Xbox 360"},
    {platform_slug::kXboxOne, "Xbox One"},
    {platform_slug::kZX80, "ZX80"},
    {platform_slug::kZX81, "ZX81"},
    {platform_slug::kZXS, "ZX Spectrum"},
};

}  // namespace

bool is_valid_slug(std::string_view slug) {
    return kSlugNames.find(slug) != kSlugNames.end();
}

std::string slug_name(std::string_view slug) {
    auto it = kSlugNames.find(slug);
    if (it != kSlugNames.end()) {
        return std::string(it->second);
    }
    return std::string(slug);
}

std::vector<std::string_view> all_slugs() {
    std::vector<std::string_view> slugs;
    slugs.reserve(kSlugNames.size());
    for (const auto& [slug, _] : kSlugNames) {
        slugs.push_back(slug);
    }
    return slugs;
}

}  // namespace retro_metadata

File: pkg/platform/slug.go
// Package platform provides platform slug constants and provider ID mappings.
package platform

// Slug represents a universal platform identifier.
// These slugs serve as common identifiers that can be mapped to
// provider-specific platform IDs (IGDB, MobyGames, ScreenScraper, etc.).
type Slug string

// Platform slug constants for cross-provider compatibility.
const (
	Slug3DO                   Slug = "3do"
	SlugAcpc                  Slug = "acpc"
	SlugAmiga                 Slug = "amiga"
	SlugAmigaCD               Slug = "amiga-cd"
	SlugAmigaCD32             Slug = "amiga-cd32"
	SlugAmstradGX4000         Slug = "amstrad-gx4000"
	SlugAndroid               Slug = "android"
	SlugAppleII               Slug = "appleii"
	SlugAppleIIGS             Slug = "apple-iigs"
	SlugArcade                Slug = "arcade"
	SlugAtari2600             Slug = "atari2600"
	SlugAtari5200             Slug = "atari5200"
	SlugAtari7800             Slug = "atari7800"
	SlugAtari8bit             Slug = "atari8bit"
	SlugAtariJaguarCD         Slug = "atari-jaguar-cd"
	SlugAtariST               Slug = "atari-st"
	SlugAtariXEGS             Slug = "atari-xegs"
	SlugBBCMicro              Slug = "bbcmicro"
	SlugC128                  Slug = "c128"
	SlugC16                   Slug = "c16"
	SlugC64                   Slug = "c64"
	SlugCPlus4                Slug = "c-plus-4"
	SlugColecovision          Slug = "colecovision"
	SlugCommodoreCDTV         Slug = "commodore-cdtv"
	SlugCPS1                  Slug = "cps1"
	SlugCPS2                  Slug = "cps2"
	SlugCPS3                  Slug = "cps3"
	SlugDC                    Slug = "dc"
	SlugDOS                   Slug = "dos"
	SlugEvercade              Slug = "evercade"
	SlugFairchildChannelF     Slug = "fairchild-channel-f"
	SlugFamicom               Slug = "famicom"
	SlugFDS                   Slug = "fds"
	SlugFMTowns               Slug = "fm-towns"
	SlugGameGear              Slug = "gamegear"
	SlugGamate                Slug = "gamate"
	SlugGameDotCom            Slug = "game-dot-com"
	SlugGB                    Slug = "gb"
	SlugGBA                   Slug = "gba"
	SlugGBC                   Slug = "gbc"
	SlugGenesis               Slug = "genesis"
	SlugGizmondo              Slug = "gizmondo"
	SlugIOS                   Slug = "ios"
	SlugIntellvision          Slug = "intellivision"
	SlugJaguar                Slug = "jaguar"
	SlugLinux                 Slug = "linux"
	SlugLynx                  Slug = "lynx"
	SlugMac                   Slug = "mac"
	SlugMSX                   Slug = "msx"
	SlugMSX2                  Slug = "msx2"
	SlugMSX2Plus              Slug = "msx2plus"
	SlugN3DS                  Slug = "3ds"
	SlugN64                   Slug = "n64"
	SlugN64DD                 Slug = "64dd"
	SlugNDS                   Slug = "nds"
	SlugNeoGeoAES             Slug = "neogeoaes"
	SlugNeoGeoCD              Slug = "neo-geo-cd"
	SlugNeoGeoMVS             Slug = "neogeomvs"
	SlugNeoGeoPocket          Slug = "neo-geo-pocket"
	SlugNeoGeoPocketColor     Slug = "neo-geo-pocket-color"
	SlugNES                   Slug = "nes"
	SlugNewNintendo3DS        Slug = "new-nintendo-3ds"
	SlugNGage                 Slug = "ngage"
	SlugNGC                   Slug = "ngc"
	SlugNintendoDSi           Slug = "nintendo-dsi"
	SlugOdyssey2              Slug = "odyssey-2"
	SlugOuya                  Slug = "ouya"
	SlugPC8800                Slug = "pc-8800-series"
	SlugPC9800                Slug = "pc-9800-series"
	SlugPCFX                  Slug = "pc-fx"
	SlugPlaydate              Slug = "playdate"
	SlugPocketstation         Slug = "pocketstation"
	SlugPokemonMini           Slug = "pokemon-mini"
	SlugPS2                   Slug = "ps2"
	SlugPS3                   Slug = "ps3"
	SlugPS4                   Slug = "ps4"
	SlugPS5                   Slug = "ps5"
	SlugPSP                   Slug = "psp"
	SlugPSVita                Slug = "psvita"
	SlugPSVR                  Slug = "psvr"
	SlugPSVR2                 Slug = "psvr2"
	SlugPSX                   Slug = "psx"
	SlugSatellaview           Slug = "satellaview"
	SlugSaturn                Slug = "saturn"
	SlugSega32                Slug = "sega32"
	SlugSegaCD                Slug = "segacd"
	SlugSegaCD32              Slug = "segacd32"
	SlugSegaPico              Slug = "sega-pico"
	SlugSeriesXS              Slug = "series-x-s"
	SlugSFam                  Slug = "sfam"
	SlugSG1000                Slug = "sg1000"
	SlugSharpX68000           Slug = "sharp-x68000"
	SlugSMS                   Slug = "sms"
	SlugSNES                  Slug = "snes"
	SlugStadia                Slug = "stadia"
	SlugSuperGrafx            Slug = "supergrafx"
	SlugSupervision           Slug = "supervision"
	SlugSwitch                Slug = "switch"
	SlugTG16                  Slug = "tg16"
	SlugTurboGrafxCD          Slug = "turbografx-cd"
	SlugVectrex               Slug = "vectrex"
	SlugVIC20                 Slug = "vic-20"
	SlugVirtualBoy            Slug = "virtualboy"
	SlugWii                   Slug = "wii"
	SlugWiiU                  Slug = "wiiu"
	SlugWin                   Slug = "win"
	SlugWin3x                 Slug = "win3x"
	SlugWonderSwan            Slug = "wonderswan"
	SlugWonderSwanColor       Slug = "wonderswan-color"
	SlugX1                    Slug = "x1"
	SlugXbox                  Slug = "xbox"
	SlugXbox360               Slug = "xbox360"
	SlugXboxOne               Slug = "xboxone"
	SlugZX80                  Slug = "zx80"
	SlugZX81                  Slug = "zx81"
	SlugZXS                   Slug = "zxs"
)

// String returns the string representation of the slug.
func (s Slug) String() string {
	return string(s)
}

// IsValid checks if the slug is a known platform.
func (s Slug) IsValid() bool {
	_, exists := slugNames[s]
	return exists
}

// Name returns the human-readable name for the platform.
func (s Slug) Name() string {
	if name, ok := slugNames[s]; ok {
		return name
	}
	return string(s)
}

// slugNames maps slugs to human-readable names.
var slugNames = map[Slug]string{
	Slug3DO:               "3DO Interactive Multiplayer",
	SlugAcpc:              "Amstrad CPC",
	SlugAmiga:             "Amiga",
	SlugAmigaCD:           "Amiga CD",
	SlugAmigaCD32:         "Amiga CD32",
	SlugAmstradGX4000:     "Amstrad GX4000",
	SlugAndroid:           "Android",
	SlugAppleII:           "Apple II",
	SlugAppleIIGS:         "Apple IIGS",
	SlugArcade:            "Arcade",
	SlugAtari2600:         "Atari 2600",
	SlugAtari5200:         "Atari 5200",
	SlugAtari7800:         "Atari 7800",
	SlugAtari8bit:         "Atari 8-bit",
	SlugAtariJaguarCD:     "Atari Jaguar CD",
	SlugAtariST:           "Atari ST",
	SlugAtariXEGS:         "Atari XEGS",
	SlugBBCMicro:          "BBC Micro",
	SlugC128:              "Commodore 128",
	SlugC16:               "Commodore 16",
	SlugC64:               "Commodore 64",
	SlugCPlus4:            "Commodore Plus/4",
	SlugColecovision:      "ColecoVision",
	SlugCommodoreCDTV:     "Commodore CDTV",
	SlugCPS1:              "CPS-1",
	SlugCPS2:              "CPS-2",
	SlugCPS3:              "CPS-3",
	SlugDC:                "Sega Dreamcast",
	SlugDOS:               "DOS",
	SlugEvercade:          "Evercade",
	SlugFairchildChannelF: "Fairchild Channel F",
	SlugFamicom:           "Famicom",
	SlugFDS:               "Famicom Disk System",
	SlugFMTowns:           "FM Towns",
	SlugGameGear:          "Sega Game Gear",
	SlugGamate:            "Gamate",
	SlugGameDotCom:        "Game.com",
	SlugGB:                "Game Boy",
	SlugGBA:               "Game Boy Advance",
	SlugGBC:               "Game Boy Color",
	SlugGenesis:           "Sega Genesis",
	SlugGizmondo:          "Gizmondo",
	SlugIOS:               "iOS",
	SlugIntellvision:      "Intellivision",
	SlugJaguar:            "Atari Jaguar",
	SlugLinux:             "Linux",
	SlugLynx:              "Atari Lynx",
	SlugMac:               "Mac",
	SlugMSX:               "MSX",
	SlugMSX2:              "MSX2",
	SlugMSX2Plus:          "MSX2+",
	SlugN3DS:              "Nintendo 3DS",
	SlugN64:               "Nintendo 64",
	SlugN64DD:             "Nintendo 64DD",
	SlugNDS:               "Nintendo DS",
	SlugNeoGeoAES:         "Neo Geo AES",
	SlugNeoGeoCD:          "Neo Geo CD",
	SlugNeoGeoMVS:         "Neo Geo MVS",
	SlugNeoGeoPocket:      "Neo Geo Pocket",
	SlugNeoGeoPocketColor: "Neo Geo Pocket Color",
	SlugNES:               "Nintendo Entertainment System",
	SlugNewNintendo3DS:    "New Nintendo 3DS",
	SlugNGage:             "N-Gage",
	SlugNGC:               "Nintendo GameCube",
	SlugNintendoDSi:       "Nintendo DSi",
	SlugOdyssey2:          "Magnavox Odyssey 2",
	SlugOuya:              "Ouya",
	SlugPC8800:            "PC-8800 Series",
	SlugPC9800:            "PC-9800 Series",
	SlugPCFX:              "PC-FX",
	SlugPlaydate:          "Playdate",
	SlugPocketstation:     "PocketStation",
	SlugPokemonMini:       "Pokemon Mini",
	SlugPS2:               "PlayStation 2",
	SlugPS3:               "PlayStation 3",
	SlugPS4:               "PlayStation 4",
	SlugPS5:               "PlayStation 5",
	SlugPSP:               "PlayStation Portable",
	SlugPSVita:            "PlayStation Vita",
	SlugPSVR:              "PlayStation VR",
	SlugPSVR2:             "PlayStation VR2",
	SlugPSX:               "PlayStation",
	SlugSatellaview:       "Satellaview",
	SlugSaturn:            "Sega Saturn",
	SlugSega32:            "Sega 32X",
	SlugSegaCD:            "Sega CD",
	SlugSegaCD32:          "Sega CD 32X",
	SlugSegaPico:          "Sega Pico",
	SlugSeriesXS:          "Xbox Series X|S",
	SlugSFam:              "Super Famicom",
	SlugSG1000:            "Sega SG-1000",
	SlugSharpX68000:       "Sharp X68000",
	SlugSMS:               "Sega Master System",
	SlugSNES:              "Super Nintendo",
	SlugStadia:            "Google Stadia",
	SlugSuperGrafx:        "SuperGrafx",
	SlugSupervision:       "Supervision",
	SlugSwitch:            "Nintendo Switch",
	SlugTG16:              "TurboGrafx-16",
	SlugTurboGrafxCD:      "TurboGrafx-CD",
	SlugVectrex:           "Vectrex",
	SlugVIC20:             "VIC-20",
	SlugVirtualBoy:        "Virtual Boy",
	SlugWii:               "Wii",
	SlugWiiU:              "Wii U",
	SlugWin:               "Windows",
	SlugWin3x:             "Windows 3.x",
	SlugWonderSwan:        "WonderSwan",
	SlugWonderSwanColor:   "WonderSwan Color",
	SlugX1:                "Sharp X1",
	SlugXbox:              "Xbox",
	SlugXbox360:           "Xbox 360",
	SlugXboxOne:           "Xbox One",
	SlugZX80:              "ZX80",
	SlugZX81:              "ZX81",
	SlugZXS:               "ZX Spectrum",
}

// AllSlugs returns all defined platform slugs.
func AllSlugs() []Slug {
	slugs := make([]Slug, 0, len(slugNames))
	for slug := range slugNames {
		slugs = append(slugs, slug)
	}
	return slugs
}
